index.js

```diff
@@ -0,0 +1,52 @@
+'use strict';
+
+const aws4 = require('aws4');
+const url = require('url');
+
+module.exports = function awsSignedRequest(service, options) {
+  if (!service || typeof service !== 'string') {
+    throw new Error('The AWS service you wish to make a request to must be provided as the first parameter. View docs for list of available services.');
+  }
+  options = options || {};
+  options.key = options.key || process.env.AWS_ACCESS_KEY_ID;
+  options.secret = options.secret || process.env.AWS_SECRET_ACCESS_KEY;
+  options.region = options.region || process.env.AWS_REGION;
+
+  return function signRequest(req) {
+    const end = req.end;
+    req.end = function(cb) {
+      let data = req.header['Content-Type'] === 'application/json'
+        ? JSON.stringify(req._data)
+        : req._formData;
+
+      const headers = req.header;
+      if (options.sessionToken) {
+        headers['X-Amz-Security-Token'] = options.sessionToken;
+      }
+
+      const parsedUrl = url.parse(req.url);
+      const signedOptions = aws4.sign({
+        host: parsedUrl.host,
+        method: req.method,
+        path: parsedUrl.path,
+        body: data,
+        service: service,
+        region: options.region,
+        headers: headers
+      }, {
+        accessKeyId: options.key,
+        secretAccessKey: options.secret,
+        sessionToken: options.sessionToken
+      });
+
+      req.header = signedOptions.headers;
+
+      req.end = end;
```

Comment:
I had seen  other plugins doing this type of thing done this way and wonder if there could be potential for another plugin that comes after this one that needs access to the original end function on the request